File: baekjoon_all/01000+/boj_1569.cpp
// Solve 2024-02-02
// Update 2025-03-04

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;

int main() {
    FASTIO;

    int n;
    cin >> n;

    int x[50], y[50];
    int min_x = 1001, max_x = -1001;
    int min_y = 1001, max_y = -1001;

    for (int i = 0; i < n; i++) {
        cin >> x[i] >> y[i];
        min_x = min(min_x, x[i]);
        max_x = max(max_x, x[i]);
        min_y = min(min_y, y[i]);
        max_y = max(max_y, y[i]);
    }

    int len_x = max_x - min_x;
    int len_y = max_y - min_y;
    int square_len = max(len_x, len_y);

    if (len_x > len_y) {
        int y1 = max_y - square_len;
        int y2 = max_y;
        bool possible = false;

        while (y1 <= min_y) {
            bool all_points_on_square = true;

            for (int i = 0; i < n; i++) {
                if (x[i] != min_x && x[i] != max_x && y[i] != y1 && y[i] != y2) {
                    all_points_on_square = false;
                    break;
                }
            }

            if (all_points_on_square) {
                possible = true;
                break;
            }

            y1++;
            y2++;
        }

        if (!possible) {
            square_len = -1;
        }
    }
    else if (len_x < len_y) {
        int x1 = max_x - square_len;
        int x2 = max_x;
        bool possible = false;

        while (x1 <= min_x) {
            bool all_points_on_square = true;

            for (int i = 0; i < n; i++) {
                if (x[i] != x1 && x[i] != x2 && y[i] != min_y && y[i] != max_y) {
                    all_points_on_square = false;
                    break;
                }
            }

            if (all_points_on_square) {
                possible = true;
                break;
            }

            x1++;
            x2++;
        }

        if (!possible) {
            square_len = -1;
        }
    }
    else {
        for (int i = 0; i < n; i++) {
            if (x[i] != min_x && x[i] != max_x && y[i] != min_y && y[i] != max_y) {
                square_len = -1;
                break;
            }
        }
    }

    cout << square_len << '\n';

    return 0;
}

File: baekjoon_all/01000+/boj_1569.java
// Solve 2024-02-02
// Update 2025-03-04

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1569 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int[] x = new int[n];
        int[] y = new int[n];

        int minX = 1001;
        int maxX = -1001;
        int minY = 1001;
        int maxY = -1001;

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            x[i] = Integer.parseInt(st.nextToken());
            y[i] = Integer.parseInt(st.nextToken());

            minX = Math.min(minX, x[i]);
            maxX = Math.max(maxX, x[i]);
            minY = Math.min(minY, y[i]);
            maxY = Math.max(maxY, y[i]);
        }

        int lenX = maxX - minX;
        int lenY = maxY - minY;
        int squareLength = Math.max(lenX, lenY);

        if (lenX > lenY) {
            int y1 = maxY - squareLength;
            int y2 = maxY;
            boolean possible = false;

            while (y1 <= minY) {
                boolean allPointsOnSquare = true;

                for (int i = 0; i < n; i++) {
                    if (x[i] != minX && x[i] != maxX && y[i] != y1 && y[i] != y2) {
                        allPointsOnSquare = false;
                        break;
                    }
                }

                if (allPointsOnSquare) {
                    possible = true;
                    break;
                }

                y1++;
                y2++;
            }

            if (!possible) {
                squareLength = -1;
            }
        } else if (lenX < lenY) {
            int x1 = maxX - squareLength;
            int x2 = maxX;
            boolean possible = false;

            while (x1 <= minX) {
                boolean allPointsOnSquare = true;

                for (int i = 0; i < n; i++) {
                    if (x[i] != x1 && x[i] != x2 && y[i] != minY && y[i] != maxY) {
                        allPointsOnSquare = false;
                        break;
                    }
                }

                if (allPointsOnSquare) {
                    possible = true;
                    break;
                }

                x1++;
                x2++;
            }

            if (!possible) {
                squareLength = -1;
            }
        } else {
            for (int i = 0; i < n; i++) {
                if (x[i] != minX && x[i] != maxX && y[i] != minY && y[i] != maxY) {
                    squareLength = -1;
                    break;
                }
            }
        }

        System.out.println(squareLength);
    }

}
